src/main/java/org/spongepowered/api/event/entity/living/player/SteerVehicleEvent.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * This file is part of SpongeAPI, licensed under the MIT License (MIT).
+ *
+ * Copyright (c) SpongePowered <https://www.spongepowered.org>
+ * Copyright (c) contributors
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event.entity.living.player;
+
+import org.spongepowered.api.entity.living.player.server.ServerPlayer;
+import org.spongepowered.api.event.Event;
+import org.spongepowered.api.util.annotation.eventgen.GenerateFactoryMethod;
+
+/**
+ * Called each tick for all {@link ServerPlayer players} riding a vehicle.
+ */
+@GenerateFactoryMethod
+public interface SteerVehicleEvent extends Event {
+
+    /**
+     * Get the {@link ServerPlayer player}.
+     *
+     * @return The player
+     */
+    ServerPlayer player();
+
+    /**
+     * Get whether the player is steering left or right.
+     *
+     * @return The sway
+     */
+    Sway sway();
+
+    /**
+     * Get whether the player is steering forwards or backwards.
+     *
+     * @return The surge
+     */
+    Surge surge();
+
+    /**
+     * Get whether the player is attempting to jump.
+     *
+     * @return Whether the player is attempting to jump.
+     */
+    boolean jump();
+
+    /**
+     * Get whether the player is attempting to dismount the entity.
+     *
+     * @return Whether the player is attempting to dismount the entity.
+     */
+    boolean unmount();
+
+    /**
+     * The possible scenarios of steering left or right.
+     */
+    enum Sway {
```

Comment:
I understand it's effectively either left, right, or none, but I'm partial to still not having any enums here.

Reply:
I’d argue that using enums is the best solution here, since it’s more user friendly than the alternatives (I can think of). Additionally, they allow you to use switch statements, which is quite nice.

The alternatives that I can think of are
a) Just providing the float values straight from the packet
b) Having 4 distinct booleans (left, right, forwards, backwards)

The problem I have with alternative a is that it’s very easy to confuse the two directions. Does a positive number mean left or right?

Alternative b is also problematic in terms of clarity. Having distinct booleans for pressing left and right may give someone the impression that left and right may be true simultaneously, if both a and d are pressed.

Of course, both of these problems can be solved with good javadoc comments. However I personally still prefer enums, because I find them more user friendly and I like switch statements :P

Reply:
The bigger problem I have here is that you're bleeding implementation details into the API.

Take this on its own, why is "RIGHT" or "LEFT" -1 and 1? Why does a plugin developer care? Quite simply, they don't and they shouldn't. If Mojang change the packet, we have to change the API in a breaking way. This is why we have our pseudo enums that hide as much implementation as possible.

If we are to have enums, they should go in the `util` package and be named in a neutral way so they can be reused.

That said, thinking beyond this, it's quite clear that the packets can return sways and surges between -1 and 1, not just 1, 0 and -1 - is this information useful? For sway, is it simply an angular velocity (so, should be expose degrees per unit)? For surge, is it a linear velocity?